Add mipi csi2 rx driver

ngphibang opened the pull request:
This PR adds mipi csi2 rx driver. It is splitted from https://github.com/zephyrproject-rtos/zephyr/pull/69810 to ease the review process.

Comments from Daniel Degrasse are addressed (at my best).

ngphibang committed:
dts: arm: nxp: Add devicetree node for MIPI CSI-2 Rx

Add a node for MIPI CSI-2 Rx in i.MX RT11xx devicetree and connect it to
the CSI node.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
soc: nxp: imxrt11xx: Enable power, clocks and muxing for MIPI CSI-2 Rx

Enable power and clocks for MIPI CSI-2 Rx.
Configure Video Mux to connect it to CSI.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Add NXP MIPI CSI-2 Rx driver

Add driver to support NXP MIPI CSI-2 Rx which is a MIPI CSI-2 receiver
connecting a camera sensor to the NXP CSI. This IP is present in the
i.MX RT11XX series.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts: bindings: video: Add bindings for NXP MIPI CSI-2 Rx

Add bindings for NXP MIPI CSI-2 Rx which is a MIPI CSI-2 receiver
connecting a camera sensor to the NXP CSI.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

danieldegrasse commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -0,0 +1,217 @@
+	return 0;
+}
+
+DEVICE_DT_INST_DEFINE(0, &mipi_csi2rx_init_0, NULL, &mipi_csi2rx_data_0,
```
Why not make this driver multi-instance? It seems like we've done all the legwork of avoiding global static variables, so initializing this driver in a multi instance macro wouldn't be a lot more effort (and would make it possible to support multiple CSI2RX peripherals if we chose to implement that on an SOC in the future)

ngphibang replied:
Yes, done. I just want to add that in the "future", even if we want to support multiple camera sensors at the same time, it does not necessarily to replicate the CSI2RX. Except for very simple IPs, we could do the replication.
But in general, we don't do that to be able to reuse things (registers, for example), so multi-instance driver seems not needed for this case, I think.

ngphibang commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -0,0 +1,217 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mipi_csi);
+
+#define DEFAULT_MIPI_CSI_NUM_LANES 2
```
yes, done

ngphibang commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -0,0 +1,217 @@
+	}
+
+	/* MIPI DPHY power on and isolation off */
+#if defined(CONFIG_SOC_SERIES_IMXRT11XX)
```
Yes, done.

ngphibang commented on dts/bindings/video/nxp,mipi-csi2rx.yaml:
```diff
@@ -0,0 +1,17 @@
+  sensor:
+    required: true
+    type: phandle
+    description: phandle of connected sensor device
```
Yes, done

ngphibang commented:
@danieldegrasse , @decsny Thanks. Is it ok for you now ?

danieldegrasse commented:
Which PR has the devicetree definition for the CSI2RX on the RT11xx after the initial PR was split? At the very least, the SOC devicetree definition can be added here (I know we may need to wait on the board definition until #71854 merges)

ngphibang commented:
@danieldegrasse : Ok. Added [the commit](https://github.com/zephyrproject-rtos/zephyr/pull/71859/commits/098c6eafff40a7bf9d2ba7d4549f743d45bc1068) that adds MIPI CSI2Rx node. BTW, it depend on https://github.com/zephyrproject-rtos/zephyr/pull/72420, not https://github.com/zephyrproject-rtos/zephyr/pull/71854

ngphibang commented on drivers/video/Kconfig.mcux_mipi_csi2rx:
```diff
@@ -0,0 +1,10 @@
+	bool "NXP MIPI CSI-2 Rx driver"
+	default y
+	depends on DT_HAS_NXP_MIPI_CSI2RX_ENABLED
+	select VIDEO_MCUX_CSI
```
Yes, but it is done in https://github.com/zephyrproject-rtos/zephyr/pull/71860

ngphibang replied:
Otherwise, this does not relate to this PR, right  ?

danieldegrasse replied:
Sorry, I missed that PR- it does relate, because the `select VIDEO_MCUX_CSI` statement here will trigger a Kconfig error on the RT11xx series when building the CSI2X driver, because `VIDEO_MCUX_CSI` currently has a dependency on `HAS_MCUX_CSI`. All this means is that this PR is dependent on #71860, as far as I can tell

ngphibang replied:
But this CSI2Rx driver doesn't build. Both CSI and CSI2Rx will build only when there are sensor connected to them and it is triggered in the sensor overlay. By default, CSI2Rx and CSI nodes are disabled (to save power).

So it is ok to approve the PR alone or there's something I need to do ?

ngphibang replied:
@danieldegrasse : So, should we move the commit https://github.com/zephyrproject-rtos/zephyr/pull/72420/commits/b226dedfbc2c4cbcf37dd1c86e85895a4660afa3 here ?

danieldegrasse replied:
I'm fine to move b226dedfbc2c4cbcf37dd1c86e85895a4660afa3 into #71860.

> But this CSI2Rx driver doesn't build. Both CSI and CSI2Rx will build only when there is a camera sensor connected to them and the build is triggered in the camera sensor overlay (another PR). By default, CSI2Rx and CSI nodes are disabled.

Maybe I'm not following here- are you saying that since the CSI2Rx driver does not build as part of this PR, there is no dependency on #71860? Generally, we want to be able to build code that is introduced with a PR as part of that PR. Finding out there is a build error later on means we have to fix something in the tree, which is less clean than getting it right the first time around

ngphibang replied:
Okay ...

- Added build test for mipi csi2rx
- Split https://github.com/zephyrproject-rtos/zephyr/pull/72623 from https://github.com/zephyrproject-rtos/zephyr/pull/72420 to avoid mutual dependency

This PR is now dependent on https://github.com/zephyrproject-rtos/zephyr/pull/72623 and https://github.com/zephyrproject-rtos/zephyr/pull/71854.

Is it ok for you now ?

danieldegrasse replied:
Yes, I think this is ok. To make sure I understand the dependency chain here:

#71859: dependent on #72623 and #71854
#71854: no dependencies
#72420: dependent on #71859 (and its dependencies)
#72434: dependent on #72420 (and its dependencies), and #71854
#72632: dependent on #72434 (and its dependencies)

Is this all accurate?

ngphibang replied:
@danieldegrasse It's nearly correct. The dependency chain are:
https://github.com/zephyrproject-rtos/zephyr/pull/72435: no dependencies
https://github.com/zephyrproject-rtos/zephyr/pull/71854: no dependencies
https://github.com/zephyrproject-rtos/zephyr/pull/72623: no dependencies
https://github.com/zephyrproject-rtos/zephyr/pull/71859: dependent on https://github.com/zephyrproject-rtos/zephyr/pull/72623 and https://github.com/zephyrproject-rtos/zephyr/pull/71854
https://github.com/zephyrproject-rtos/zephyr/pull/72420: dependent on https://github.com/zephyrproject-rtos/zephyr/pull/71859 (and its dependencies).
https://github.com/zephyrproject-rtos/zephyr/pull/72434: dependent on https://github.com/zephyrproject-rtos/zephyr/pull/71854 and https://github.com/zephyrproject-rtos/zephyr/pull/71859 (and its dependency)
https://github.com/zephyrproject-rtos/zephyr/pull/72632: dependent on https://github.com/zephyrproject-rtos/zephyr/pull/72434 (and its dependencies)
https://github.com/zephyrproject-rtos/zephyr/pull/72633: dependent on https://github.com/zephyrproject-rtos/zephyr/pull/71859 (and its dependencies) and https://github.com/zephyrproject-rtos/zephyr/pull/72434 (and its dependency)

ngphibang replied:
Forgot : https://github.com/zephyrproject-rtos/zephyr/pull/72631: no dependencies

ngphibang committed:
tests: drivers: build_all: video: Add test for mipi csi2rx

Add test case for mipi csi2rx driver. As this driver selects the csi
driver, the csi node has to be enabled as well.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
Updated "sdev" => "source" according the change in https://github.com/zephyrproject-rtos/zephyr/pull/72623 (which is merged)

loicpoulain reviewed (dismissed):
LGTM.

danieldegrasse commented:
@ngphibang can you rebase this PR now that the dependencies are merged?

mmahadevan108 commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -0,0 +1,218 @@
+ * more common and more performant but single lane is also supported.
+ */
+#define DEFAULT_MIPI_CSI_NUM_LANES 2
+#define DEFAULT_CAMERA_FRAME_RATE  30
```
Is there a value to making this a Kconfig. Write now the user has to edit this driver file to change the default frame rate.

ngphibang commented:
@danieldegrasse yes, rebased now. Could you approve ?

Then, I will need @loicpoulain to approve it again (sorry for this inconvience... I couldn't rebase it earlier as yesterday is a bank holiday in France).
